packages/frontend/src/pages/Contribute.jsx

```diff
@@ -70,18 +113,110 @@ export default observer(() => {
+      const renderer = new THREE.WebGLRenderer({ antialias: true, canvas })
 
-      const ctx = canvas.getContext('2d')
-      ctx.font = '2rem Azeret Mono'
-      ctx.fillStyle = 'white'
-      ctx.textAlign = 'center'
-      ctx.fillText(
-        "Bhargov's cosmo should be here.",
-        canvas.width / 2,
-        canvas.height / 2
+      // set camera
+      let camera = new THREE.PerspectiveCamera(
+        // fov
+        35,
+        // aspect
+        width / height,
+        // near
+        0.01,
+        // far
+        10000 * AU
       )
+      const defaultZoom = 1.7 * AU
+      camera.position.set(0, 0, defaultZoom)
+
+      const scene = new THREE.Scene()
+      let clock = new THREE.Clock()
+
+      // create particle buffers
+      const particleGeometry = new THREE.BufferGeometry()
+      let blackHoleCount = 0
+      let positions = new Float32Array(3 * N_plummer)
+      let mass = new Float32Array(N_particles).fill(1, 0, N_plummer)
+      particleGeometry.setAttribute(
+        'position',
+        new THREE.BufferAttribute(positions, 3)
+      )
+      particleGeometry.setAttribute('mass', new THREE.BufferAttribute(mass, 1))
+
+      // create plane
+      const planeGeometry = new THREE.PlaneGeometry(AU, AU, 8, 8)
+      const planeMaterial = new THREE.MeshBasicMaterial({
+        color: 0x049ef4,
+        wireframe: true,
+        transparent: true,
+        opacity: 0.1,
+        side: THREE.DoubleSide,
+      })
+
+      // for particle colors
+      const particleShader = new THREE.ShaderMaterial({
+        vertexShader: vertexShaderSrc,
+        fragmentShader: fragmentShaderSrc,
+        uniforms: {},
+      })
+
+      const cameraControls = new OrbitControls(camera, renderer.domElement)
+      cameraControls.noPan = false
+
+      const vFOV = THREE.MathUtils.degToRad(camera.fov) // convert vertical fov to radians
+      const scene_height = 2 * Math.tan(vFOV / 2) * camera.position.z // visible height
+      const scene_width = scene_height * camera.aspect
+      const boundingRect = canvas.getBoundingClientRect()
+
+      canvas.addEventListener('mouseup', (e) => {
+        if (!e.shiftKey) return
+        const _x = e.clientX - boundingRect.left - width / 2
+        const _y = e.clientY - boundingRect.top - height / 2
+        const px =
+          scene_width *
+          (_x / width) *
+          (1 + Math.abs(Math.sin(cameraControls.getAzimuthalAngle())))
+        const py =
+          -scene_height *
+          (_y / height) *
+          (1 + Math.abs(Math.cos(cameraControls.getPolarAngle())))
+        fmm.insert_particle(px, py, 0, 1e29)
+        blackHoleCount += 1
+        mass = new Float32Array(mass.length + 1)
+          .fill(1, 0, N_plummer)
+          .fill(10, N_particles, mass.length + 1)
+        particleGeometry.setAttribute(
+          'mass',
+          new THREE.BufferAttribute(mass, 1)
+        )
+        particleGeometry.attributes.mass.needsUpdate = true
+      })
```

Comment:
I think the position of the star could be better. Under this algorithm, sometimes under some angle I won't find my star until it runs to me.
Also while adding a new star, maybe the mass could range from `2~20` or something else instead of just 10?

Reply:
I agree that this is a very crude algorithm for placing stars, and I'll have to work the math out. The idea was that a star should appear exactly where the user clicks a part of the screen while looking along the z-axis into the plane of the galaxy. However, if the camera is tilted, the position where the user clicks should be projected onto the plane of the galaxy.